File: core/src/main/clojure/xtdb/tracer.clj
(ns xtdb.tracer
  (:require [xtdb.node :as xtn])
  (:import (xtdb.api Xtdb$Config)
           (xtdb.api.metrics TracerConfig)))

(defmethod xtn/apply-config! :xtdb/tracer [^Xtdb$Config config _ {:keys [enabled? endpoint service-name query-tracing? transaction-tracing? span-processor]}]
  (.tracer config
           (cond-> (TracerConfig.)
             (some? enabled?) (.enabled enabled?)
             endpoint (.endpoint endpoint)
             service-name (.serviceName service-name)
             (some? query-tracing?) (.queryTracing query-tracing?)
             (some? transaction-tracing?) (.transactionTracing transaction-tracing?)
             span-processor (.spanProcessor span-processor))))

File: core/src/main/kotlin/xtdb/Tracer.kt
package xtdb

import io.micrometer.tracing.otel.bridge.OtelCurrentTraceContext
import io.micrometer.tracing.otel.bridge.OtelTracer
import io.opentelemetry.api.common.Attributes
import io.opentelemetry.exporter.otlp.http.trace.OtlpHttpSpanExporter
import io.opentelemetry.sdk.OpenTelemetrySdk
import io.opentelemetry.sdk.resources.Resource
import io.opentelemetry.sdk.trace.SdkTracerProvider
import io.opentelemetry.sdk.trace.export.SimpleSpanProcessor
import xtdb.api.metrics.TracerConfig
import xtdb.util.info
import xtdb.util.logger
import xtdb.util.warn

object Tracer {
    private val LOG = Tracer::class.logger

    @JvmStatic
    @JvmName("open")
    fun TracerConfig.openTracer(): OtelTracer? = try {
        val resource =
            Resource.getDefault().merge(Resource.create(Attributes.builder().put("service.name", serviceName).build()))

        val spanProcessor = spanProcessor
            ?: SimpleSpanProcessor.create(OtlpHttpSpanExporter.builder().setEndpoint(endpoint).build())

        val tracerProvider =
            SdkTracerProvider.builder()
                .setResource(resource)
                .addSpanProcessor(spanProcessor)
                .build()

        val tracer =
            OpenTelemetrySdk.builder()
                .setTracerProvider(tracerProvider)
                .build()
                .getTracer("xtdb")

        LOG.info("OpenTelemetry tracer created for service: $serviceName")

        OtelTracer(tracer, OtelCurrentTraceContext()) { }
    } catch (e: Throwable) {
        LOG.warn(e, "Failed to create OpenTelemetry tracer, tracing will be disabled")
        null
    }
}
